### ISSUE TITLE
Make ValueObservationScheduler work with arbitrary schedulers

### ISSUE THREAD

**Issue by ZevEisenberg** [groue/GRDB.swift#1057]
### What did you do?

I want to use my own Combine schedulers with functions like `publisher(in:scheduling:)` with my own Combine schedulers so I can do things like pass in immediate or test schedulers from The Composable Architecture. However, this function takes a `ValueObservationScheduler`, which uses dispatch queues in its public interface and private implementation details. This makes it hard to control the scheduler for testing or other uses.

### What did you expect to happen?

An API exists to construct a `ValueObservationScheduler` using an arbitrary conformer of `Combine.Scheduler`.

### What happened instead?

No such API exists, so I have to use a dispatch queue directly, which makes controlling dependencies harder.

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.8.0
**Installation method:** SPM
**Xcode version:** 12.5
**Swift version:** 5.4
**Platform(s) running GRDB:** iPadOS
**macOS version running Xcode:** macOS 11.5.2 (20G95)

### Demo Project

N/A

**Comment by groue** [comment]
Hello @ZevEisenberg,

I'm listening, and at the same time I'm not sure I can correctly answer your request.

First, Combine support is a convenience layer. ValueObservation does not depend on Combine. It works on systems that do not ship with Combine.

Next, ValueObservation notifications are scheduled by GRDB, but only as a consequence of SQLite events.

Eventually, you need to be more precise in the description of your needs. And also why does the standard `receive(on:)` Combine operator is not enough for you.

**Comment by ZevEisenberg** [comment]
@groue thanks for your thoughtful reply! I'm pretty new to GRDB, and I'm joining a project that's already using it, so please forgive my ignorance. I will talk with the originator of the code base and see if I'm missing something, or if there's information I can add to the request.

**Comment by groue** [comment]
Hello @ZevEisenberg, do you have any update?

**Comment by ZevEisenberg** [comment]
Here's some code in my app:

```swift
observeExplorePage: { date in
    database
        .observe(newestBefore: date, scheduling: .async(onQueue: .global(qos: .userInitiated)))
        .removeDuplicates()
        .ignoreFailureInProduction(operation: "observeFavoriteHowTos")
        .receive(on: mainQueue)
        .eraseToEffect()
},
```

This app is using TCA, which provides some cool implementations of the `Scheduler` protocol from the `Combine` framework. `DisaptchQueue` conforms to `Scheduler`, and the `mainQueue` you see in my code sample is actually an `AnySchedulerOf<DispatchQueue>`. This lets me pass in an `ImmediateScheduler` or `TestScheduler` in my unit tests, so I can control the flow of time, make async things synchronous, and do other tricks that make tests more reliable and faster to run.

I believe the `receive(on:)` operator doesn't work in this case because, even though I can control which scheduler the stream is ultimately received on, it's still _running_ on a global dispatch queue, whereas I'd like to be able to pass in a scheduler that I fully control here:

```swift
.observe(newestBefore: date, scheduling: .async(onQueue: .global(qos: .userInitiated)))
```

So what I'd like to see if a variant of the `async(onQueue:)` function like `async(onScheduler:)` (or coalesce them both to the shared name `async(on:)`) when `#if canImport(Combine)` is true. I'm not sure what it would take to implement such a variant, but my quick read of the situation tells me that that's what I would need to be able to control this dependency, in Pointfree/TCA parlance.

This is definitely more complex than I realized when I originally posted it, since I forgot that `Scheduler` is specifically part of the Combine framework, even though `DispatchQueue` conforms to it.

**Comment by groue** [comment]
> I believe the `receive(on:)` operator doesn't work in this case because, even though I can control which scheduler the stream is ultimately received on, it's still _running_ on a global dispatch queue

That "it" which is running is the GRDB runtime, i guess. It does indeed run in GRDB-controlled dispatch queues. That's how the correctness of ValueObservation outputs is guaranteed. You can not control this runtime - that's the core of ValueObservation.

In other words, ValueObservation is a *black box*. You can only control the scheduling of *published values*, through the `scheduling` argument (`.immediate` and `.async(onQueue:)`), or the Combine `receive(on:)` operator that accept any Combine scheduler.

> So what I'd like to see if a variant of the `async(onQueue:)` function like `async(onScheduler:)` (or coalesce them both to the shared name `async(on:)`) when `#if canImport(Combine)` is true.

I understand the request, but I fail to see what you would gain, *which problem this would solve*. At best, you'd avoid a dispatch queue hop, which is not something I'd call significant. And you would not avoid the ValueObservation runtime anyway.

You may try to modify GRDB, and implement your request by adding one `ValueObservationSchedulerImpl` based on a Combine scheduler (https://github.com/groue/GRDB.swift/blob/v5.11.0/GRDB/ValueObservation/ValueObservationScheduler.swift#L64-L67). This should be trivial. But, according to my understanding, this won't solve anything.

I'm afraid we're about to talk past each other. There's a [XY problem](https://en.wikipedia.org/wiki/XY_problem) smell in this conversation. Please, instead of describing what you want, tell the reason why, or the problem you want to avoid.

**Comment by ZevEisenberg** [comment]
You're right to be wary of the XY problem, and I'm probably falling prey to it due to my inexperience with both GRDB and TCA 😅. I can only thank you for your continued patience.

I _think_ avoiding the queue hop is still useful in terms of writing a unit test that runs synchronously, and therefore is simpler to write and reason about without having to write async expectations in the test. The Composable Architecture comes with an `ImmediateScheduler` that is useful for such tasks.

But, given that we're on the eve of being able to use async/await for lots of this stuff anyway, that may be of limited utility in the medium-to-long-term. I've already taken more of your time than I deserve on this question, so feel free to close it and maybe keep an eye out for TCA-related requests. I'll report back if and when I realize that what I'm asking for makes no sense, which is likely 😆
